src/state/slices/txHistorySlice/txHistorySlice.ts

```diff
@@ -10,72 +10,66 @@ import { KnownChainIds } from '@shapeshiftoss/types'
 import orderBy from 'lodash/orderBy'
 import { getChainAdapterManager } from 'context/PluginProvider/chainAdapterSingleton'
 import { logger } from 'lib/logger'
+import type { PartialRecord } from 'lib/utils'
+import { deepUpsertArray, isSome } from 'lib/utils'
 import { BASE_RTK_CREATE_API_CONFIG } from 'state/apis/const'
 import { getFoxyApi } from 'state/apis/foxy/foxyApiSingleton'
+import type { Nominal } from 'types/common'
 
-import { addToIndex, getRelatedAssetIds, serializeTxIndex, UNIQUE_TX_ID_DELIMITER } from './utils'
+import { getRelatedAssetIds, serializeTxIndex, UNIQUE_TX_ID_DELIMITER } from './utils'
 
 const moduleLogger = logger.child({ namespace: ['txHistorySlice'] })
 
-export type TxId = string
+export type TxId = Nominal<string, 'TxId'>
 export type Tx = Transaction & { accountType?: UtxoAccountType }
 
 export type TxHistoryById = {
   [k: TxId]: Tx
 }
 
-/* this is a one to many relationship of asset id to tx id
+/* this is a one to many relationship of an account and asset id to tx id
  *
  * e.g. an account with a single trade of FOX to USDC will produce the following
  * three related assets
  *
  * {
- *   foxAssetId: [txid] // sell asset
- *   usdcAssetId: [txid] // buy asset
- *   ethAssetId: [txid] // fee asset
+ *   0xfoobaraccount: {
+ *     foxAssetId: [txid] // sell asset
+ *     usdcAssetId: [txid] // buy asset
+ *     ethAssetId: [txid] // fee asset
+ *   }
  * }
  *
  * where txid is the same txid related to all the above assets, as the
  * sell asset, buy asset, and fee asset respectively
  *
  * this allows us to O(1) select all related transactions to a given asset
+ *
+ * note - we persist this data structure to disk across page refresh, and wallet disconnections
+ * and use the accountIds from the connected wallet to index into it
  */
 
-export type TxIdByAssetId = {
-  [k: AssetId]: TxId[]
-}
-
-export type TxIdByAccountId = {
-  [k: AccountId]: TxId[]
-}
+export type TxIdsByAssetId = PartialRecord<AssetId, TxId[]>
+export type TxIdsByAccountIdAssetId = PartialRecord<AccountId, TxIdsByAssetId>
 
 // status is loading until all tx history is fetched
 export type TxHistoryStatus = 'loading' | 'loaded'
 
-type RebaseId = string
-type RebaseById = {
-  [k: RebaseId]: RebaseHistory
-}
-
-type RebaseByAssetId = {
-  [k: AssetId]: RebaseId[]
-}
+export type RebaseId = Nominal<string, 'RebaseId'>
+type RebaseById = PartialRecord<RebaseId, RebaseHistory>
 
-type RebaseByAccountId = {
-  [k: AccountId]: RebaseId[]
-}
+type RebaseIdsByAssetId = PartialRecord<AssetId, RebaseId[]>
+export type RebaseIdsByAccountIdAssetId = PartialRecord<AccountId, RebaseIdsByAssetId>
 
 export type TxsState = {
   byId: TxHistoryById
-  byAssetId: TxIdByAssetId
-  byAccountId: TxIdByAccountId
+  byAccountIdAssetId: TxIdsByAccountIdAssetId
```

Comment:
guts of the PR right here - we now index txs by `accountId` at the first level, and `assetId` at the second level. this is a minimal and exhaustive representation that we need to be able to select

* all transactions from a users accounts
* all txs for a given account
* all txs for a given asset, across multiple accounts
* txs related to a specific asset and account

`byId` is still an O(1) lookup for all txs, for all wallets that have ever been connected - and `ids` is still a sorted list of all txs ids for all wallets ever connected.